### DISCUSSION TITLE
Component composed by components

### DISCUSSION THREAD

**Discussion by Tonaion02** [bevyengine/bevy#14513] [Q&A]
What happened if i insert a Component A in an Entity where A is a Component that is composed by component B and component C?
Can i make something like that?
Can i retrieve component B or C from queries?

**Comment by piedoom** [comment]
I'm not sure if I'm totally understanding, so let me know if I'm on the wrong path, but generally not a pattern to compose components into another component - you would instead compose entities from components. What are you trying to accomplish in your app/game?

**Reply by Tonaion02** [reply]
I want to understand this:
If i have a Component A that contains two field, that are components too, can i retrieve the two contained component from the query?

**Reply by piedoom** [reply]
I see - no, you can't do that. You can technically have components in another component's fields, but it'll just be regular data and you won't be able to query it. I don't know your underlying case, but I think this might be an antipattern.

Instead, let's say you have component X and Y. To achieve this combination of components idiomatically, you can attach them to an entity with `commands.spawn((X::default(), Y::default())`. If you want to retrieve only entities that have both component X and Y, you can do that in a query, like `Query<(&X, &Y)>`. You can think of this Entity as the `Component A` in your comment, with a few differences:

Entities do not describe their data structure like a struct. The closest we can come to this is creating our own `Bundle`, but in general, strive to operate on the least amount of information possible. Instead of a master struct containing several components, we have an Entity that may contain combinations of components that have interesting interactions, which we can query with our `Query<(&X, &Y)>`. If an entity with both the X and Y component exists, it'll be returned in this query.

I hope this is helpful, feel free to drop the concrete problem you're solving too, if you'd like any opinions on idiomatic/alternative approaches!
